util-core/src/main/scala/com/twitter/util/Batcher.scala

```diff
@@ -0,0 +1,24 @@
+package com.twitter.util
+
+/** Provides a clean, lightweight interface for working with a batched [[com.twitter.util.Future]] */
+class Batcher[In, Out] private[util](
+  executor: BatchExecutor[In, Out]
+)(
+  implicit timer: Timer
+) extends Function1[In, Future[Out]] { batcher =>
+  /**
+   * Enqueues requests for a batched [[com.twitter.util.Future]]
+   *
+   * Processes all enqueued requests after the batch size is reached
+   */
+  def apply(t: In): Future[Out] = executor.enqueue(t)
+
+  /** Immediately processes all unprocessed requests */
+  def flushBatch(): Unit = {
```

Comment:
if there are guarantees regarding ordering and such, they should be documented. its not obvious from reading the diff why you synchronize on executor here.

and if there are guarantees, what prevents others callers from using the executor incorrectly?